data_warehouse/services/coverage_statistics_services.py

```diff
@@ -0,0 +1,194 @@
+'''培训覆盖率服务'''
+from django.utils.timezone import now
+from django.db.models import Q
+from django.core.exceptions import ObjectDoesNotExist
+from training_record.models import Record
+from training_program.models import Program
+from auth.models import User, Department
+from infra.exceptions import BadRequest
+
+
+class CoverageStatisticsService:
+    '''培训覆盖率统计服务'''
+    TITLES = (
+        '教授', '副教授', '讲师', '助教', '教授级工程师',
+        '高级工程师', '工程师', '研究员', '副研究员', '助理研究员')
+    AGES = (
+        ('35岁及以下', [0, 35]), ('36-45岁', [36, 45]),
+        ('46-55岁', [46, 55]), ('56岁及以上', [56, 150]))
+
+    @staticmethod
+    def teacher_coverage():
+        '''专任教师培训覆盖率统计'''
+
+    # pylint: disable=R0914
+    @staticmethod
+    def groupby_training_records(records):
```

Comment:
I know your use case ensures that you will use all of the grouped data, but in @oosprey 's case, only one of these data will be used, please take the suggestions I suggested before, construct the final data structure in the 'root' function, and the 'group by' functions should be separated into 3 functions called by the 'root' function, each of them take users (Queryset of User) as input, and return the dict of grouped statistics. BTW, your current function has significant overhead, cuz you retrieve a user instance in each forloop iteration, instead, you have `User.objects.filter(id__in=[1, 2, 3])` to retrieve users in one query. And since all 'group by' functions will use the same user instances, you should done this in the 'root' function.